cpp/ql/src/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -0,0 +1,484 @@
+import cpp
+import semmle.code.cpp.ir.IR
+
+/**
+ * A Boolean condition in the AST that guards one or more basic blocks. This includes
+ * operands of logical operators but not switch statements.
+ */
+class GuardCondition extends Expr {
+  GuardCondition() {
+    exists(IRGuardCondition ir | this = ir.getUnconvertedResultExpression())
+    or
+    // no binary operators in the IR
+    exists(Instruction ir |
+      this.(BinaryLogicalOperation).getAnOperand().getFullyConverted() = ir.getAST()
```

Comment:
Should `Instruction ir` here be `IRGuardCondition ir`? Otherwise I don't understand this case. Also, should there be a condition corresponding to the `not exists ...` in the `!x` case? Please add a comment about what that condition is for. If it's for filtering out the `!x` in `y = !x;` then don't we also want to filter out the `a && b` in `y = a && b;`?

Reply:
That is to avoid having the same expression be both a `GuardConditionFromIR` and a `GuardConditionFromShortCircuitNot`, as in 
```
y = !x;
if(y) {
 ...
}
```